more_itertools/more.py

```diff
@@ -1106,3 +1108,83 @@ def always_iterable(obj):
         return obj,
 
     return obj
+
+def adjacent(predicate, iterable, distance=1):
+    """
+    Returns an iterable over ``(bool, item)`` tuples where the ``item`` is drawn
+    from the underlying *iterable* and the ``bool`` indicates whether that item
+    satisfies the *predicate* or is adjacent to one that does. For example:
+
+        >>> list(adjacent(lambda x: x % 4 == 2, range(6)))
+        [(False, 0), (True, 1), (True, 2), (True, 3), (False, 4), (False, 5)]
+
+    The *distance* used to determine what counts as adjacent can be configured:
+
+        >>> list(adjacent(lambda x: x % 4 == 2, range(6), 2))
+        [(True, 0), (True, 1), (True, 2), (True, 3), (True, 4), (False, 5)]
+
+    In this case any elements within 2 positions of the one that satisfies the
+    predicate are paired with ``True``. *distance* can be any nonnegative
+    integer. If *distance* is zero, this reduces to ``map()`` and ``zip()``:
+
+        >>> predicate = lambda x: x % 4 == 2
+        >>> iterable = range(1000)
+        >>> with_adjacent    = adjacent(predicate, iterable, 0)
+        >>> without_adjacent = zip(map(predicate, iterable), iterable)
+        >>> all(a == b for a, b in zip(without_adjacent, with_adjacent))
+        True
+
+    This tool could be used to render contextual diffs, for example.
+
+    ``adjacent()`` is designed to avoid calling *predicate* more than once per
+    item in the *iterable*, and thus is suitable for computationally expensive
+    *predicate*s.
+
+    To group consecutive elements which are associated with the same boolean
+    value, see ``groupby_transform()``.
+    """
+    # Allow distance=0 mainly for testing that it reproduces results with map()
+    if distance < 0:
+        raise ValueError('distance must be at least 0')
+
+    i1, i2 = tee(iterable)
+    padding = [False] * distance
+    selected = chain(padding, map(predicate, i1), padding)
+    adjacent = map(any, windowed(selected, 2 * distance + 1))
+    return zip(adjacent, i2)
+
+def groupby_transform(iterable, keyfunc=None, valuefunc=None):
+    """
+    A wrapper for ``itertools.groupby()`` which computes the values in the group
+    by applying *valuefunc* to the items in the iterable.
```

Comment:
I suppose one _could_ do this:
```python
((k, map(valuefunc or lambda x: x, g) for k, g in groupby(iterable, keyfunc))
```
but that seems practically code-golfed to me ;-) It definitely does get simpler without special-casing the input arguments, though.

Do you specifically prefer generator syntax over `map()` for applying `valuefunc` to the group? Since we're using six, it's an iterable map either way, not the Python 2 `map()` which creates a list.

Since this is a pretty significant change to that one function, I'd consider doing the change on a separate branch, so the development history isn't cluttered with this version of `groupby_transform` that uses `itemgetter()`. Any preference on whether I do that, or just tack the change on to this branch?